### DISCUSSION TITLE
Exploring loosening the version compatibility check

### DISCUSSION THREAD

**Discussion by gsjaardema** [CGNS/CGNS#752] [CGNStalk]
We have some codes that are using CGNS-4.40 to read/write cgns files.  However, some of our customers want to use some visualization software that is using CGNS-3.3X.  When the customer tries to input the file to this visualizer, there is a version mismatch error.

As far as I can remember, the only(?) file format change is related to the storage of the n-gon faces/elements in 4.40 vs 3.3X, (The n-gon proposal is:  CPEX 41 NGON modification proposal (CGNS-121)) so in theory, an older cgns library version should be able to process any newer cgns library version file as long as the file is not using n-gon elements.

What are the disadvantages of modifying the version compatibility check such that a 3.3X library can read 4.X files.  Probaby output a warning message saying that if the file contains NGON elements, bad things will happen, or some other notice that this is not the recommended usage, but is being allowed under certain conditions.

I understand that applications that write files may want to stay back at earler versions so they don't break their clients who haven't updated to more recent versions, but there is also the need to be able to move up to more recent versions with new features and bug fixes...

Thanks for any and all discussion.

**Comment by gsjaardema** [comment]
I do realize that one issue with this is that it requires a modification to the 3.3X libraries which means those clients would need to update their libraries and would require another 3.3X release.   

Another option would be a setting on 4.40 and later that would allow them to specify the version as "3.X compatible" somehow, but the only way I can see to do that is to "lie" about the version number that is written to the file...

**Reply by MicK7** [reply]
If we have a mean to ensure that the 4.4 version is writing a file with a set of features fully compatible with 3.X it would be an acceptable lie. As you mentioned we can check if polyhedral cells are present (but it is not generic and the next feature introduced in the standard library may also break everything) and decide to write or not the file while lying.

The user of 4.4 library would do something like
```
cg_set_compatibility_features(3.3);
...
cg_write(...); //OK
...
cg_poly_section_write(...); // Warning appears to tell user it is doing incompatible things .

```
We would need a macro or some way to tell for each function the minimal version compatiblity.
This compatibility check should be compilation configurable to deactivate it for performance.
